extmod/uasyncio/stream.py

```diff
@@ -40,7 +40,15 @@ async def readline(self):
                 return l
 
     def write(self, buf):
-        self.out_buf += buf
+        if self.out_buf == b'':
+            ret = self.s.write(buf)
+            if ret is None:
+                ret = 0
+            elif ret == len(buf):
+                return
+            self.out_buf = buf[ret:] # must make a copy
```

Comment:
would be good to optimise the case that `ret==0`, then just store `buf`, no need to copy it

Reply:
I didn't because of the semantics of write.